src/cls/2pc_queue/cls_2pc_queue.cc

```diff
@@ -0,0 +1,433 @@
+    decode(res_op, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_2pc_queue_reserve: failed to decode entry: %s", err.what());
+    return -EINVAL;
+  }
+
+  if (res_op.size == 0) {
+    CLS_LOG(1, "ERROR: cls_2pc_queue_reserve: cannot reserve zero bytes");
+    return -EINVAL;
+  }
+  if (res_op.entries == 0) {
+    CLS_LOG(1, "ERROR: cls_2pc_queue_reserve: cannot reserve zero entries");
+    return -EINVAL;
+  }
+
+  // get head
+  cls_queue_head head;
+  int ret = queue_read_head(hctx, head);
+  if (ret < 0) {
+    return ret;
+  }
+
+  cls_2pc_urgent_data urgent_data;
+  try {
+    auto in_iter = head.bl_urgent_data.cbegin();
+    decode(urgent_data, in_iter);
+  } catch (buffer::error& err) {
+    CLS_LOG(1, "ERROR: cls_2pc_queue_reserve: failed to decode entry: %s", err.what());
+    return -EINVAL;
+  }
+
+  const auto overhead = res_op.entries*QUEUE_ENTRY_OVERHEAD;
+  const auto remaining_size = (head.tail.offset >= head.front.offset) ?
+    (head.queue_size - head.tail.offset) + (head.front.offset - head.max_head_size) :
+    head.front.offset - head.tail.offset;  
+
+
+  if (res_op.size + urgent_data.reserved_size + overhead > remaining_size) {
+    CLS_LOG(1, "ERROR: cls_2pc_queue_reserve: reservations exceeded maximum capacity");
+    CLS_LOG(10, "INFO: cls_2pc_queue_reserve: remaining size: %lu (bytes)", remaining_size);
+    CLS_LOG(10, "INFO: cls_2pc_queue_reserve: current reservations: %lu (bytes)", urgent_data.reserved_size);
+    CLS_LOG(10, "INFO: cls_2pc_queue_reserve: requested size: %lu (bytes)", res_op.size);
+    return -ENOSPC;
+  }
+
+  urgent_data.reserved_size += res_op.size + overhead;
+  // note that last id is incremented regadless of failures
+  // to avoid "old reservation" issues below
+  ++urgent_data.last_id;
+  bool result;
+  cls_2pc_reservations::iterator last_reservation;
+  std::tie(last_reservation, result) = urgent_data.reservations.emplace(std::piecewise_construct,
+          std::forward_as_tuple(urgent_data.last_id),
+          std::forward_as_tuple(res_op.size, ceph::real_clock::now()));
+  if (!result) {
+    // an old reservation that was never committed or aborted is in the map
+    // caller should try again assuming other IDs are ok
+    CLS_LOG(1, "ERROR: cls_2pc_queue_reserve: reservation id conflict after rollover: %u", urgent_data.last_id);
+    return -EAGAIN;
+  }
+
+  // write back head
+  head.bl_urgent_data.clear();
+  encode(urgent_data, head.bl_urgent_data);
+
+  const uint64_t urgent_data_length = head.bl_urgent_data.length();
+  auto total_entries = urgent_data.reservations.size();
+
+  if (head.max_urgent_data_size < urgent_data_length) {
+    CLS_LOG(10, "INFO: cls_2pc_queue_reserve: urgent data size: %lu exceeded maximum: %lu using xattrs", urgent_data_length, head.max_urgent_data_size);
+    // add the last reservation to xattrs 
+    bufferlist bl_xattrs;
+    auto ret = cls_cxx_getxattr(hctx, CLS_QUEUE_URGENT_DATA_XATTR_NAME, &bl_xattrs);
```

Comment:
this xattr spillover is duplicated in cls_rgw_gc too, is there anything significantly different about what you do here?

i feel like it should be cls_queue's responsibility to provide interfaces for opaque urgent data that handles the spillover cases transparently

many ops don't need to read/write urgent data, so we could leave `queue_read/write_head()` unchanged, and add new calls for the ops that do - for example, a `queue_write_head_urgent()` would check whether `head.urgent_data` was too big, and split the remainder into a separate bufferlist for `cls_cxx_setxattr()`. similarly, a `queue_read_head_urgent()` could check a new `head.has_xattrs` flag and conditionally call `cls_cxx_getxattr()` to splice that data onto the end of `head.urgent_data`

that way, callers like cls_2pc_queue and cls_rgw_gc could encode/decode `head.urgent_data` without worrying about the spillover themselves. and ops that don't care about urgent data would continue calling `queue_read_head()` (which only sees the part of `head.urgent_data` actually stored in the head) and `queue_write_head()` would write it back unmodified

Reply:
Seems reasonable to me.  Not quite relatedly, we need to remember that the purpose of the queuing ops is to avoid kv-store operations, so the existence of urgent data assumes that it's going to be used infrequently.  We want to avoid accidentally landing in a space where (e.g., in this case) notification intent logging runs the queue full and constantly updates xattrs.

Reply:
> Seems reasonable to me. Not quite relatedly, we need to remember that the purpose of the queuing ops is to avoid kv-store operations, so the existence of urgent data assumes that it's going to be used infrequently. We want to avoid accidentally landing in a space where (e.g., in this case) notification intent logging runs the queue full and constantly updates xattrs.

urgent data is used almost in every operation, but should, usually, not spillover to be using xattrs.
when the queue is full, we return an error, and don't use xattrs. it is only when we run out of space in the head area that we write our queue bookkeeping there.

Reply:
> this xattr spillover is duplicated in cls_rgw_gc too, is there anything significantly different about what you do here?
> 
> i feel like it should be cls_queue's responsibility to provide interfaces for opaque urgent data that handles the spillover cases transparently
> 
> many ops don't need to read/write urgent data, so we could leave `queue_read/write_head()` unchanged, and add new calls for the ops that do - for example, a `queue_write_head_urgent()` would check whether `head.urgent_data` was too big, and split the remainder into a separate bufferlist for `cls_cxx_setxattr()`. similarly, a `queue_read_head_urgent()` could check a new `head.has_xattrs` flag and conditionally call `cls_cxx_getxattr()` to splice that data onto the end of `head.urgent_data`
> 
> that way, callers like cls_2pc_queue and cls_rgw_gc could encode/decode `head.urgent_data` without worrying about the spillover themselves. and ops that don't care about urgent data would continue calling `queue_read_head()` (which only sees the part of `head.urgent_data` actually stored in the head) and `queue_write_head()` would write it back unmodified

there are two issues with that:
* current implementation does not split the buffer list. it does that at the payload level:
if the urgent data head exceed the max, we remove the entry from the actual container, and add that to a new container in xattrs.
making that work at bufferlist level would require some changes to the code
* another issue would be that, in some cases, the decision on reading from xattrs is not only related to whether xattrs exists or not, but also to whether a matching entry was found. the the value being searched and the container to search in are application dependent, this would be difficult to do that in the `cls_queue`

both could be addressed, but would probably require more complex APIs. would prefer not to do that as part of this review